File: ios/BiblosSurferTests/Support/TestFixtures.swift
//
//  TestFixtures.swift
//  BiblosSurferTests
//

import Foundation
import ReadiumShared
@testable import BiblosSurfer

enum TestFixtures {
    static let sampleBookTitle = UITestStubLibraryService.sampleTitle
    static let sampleBookAuthor = UITestStubLibraryService.sampleAuthor

    /// The EPUB bundled with the app target (Wolne Lektury Genesis / Cylkow).
    /// Unit tests run inside the app host, so `Bundle.main` resolves it.
    static var sampleBookURL: URL? {
        Bundle.main.url(
            forResource: UITestStubLibraryService.resourceName,
            withExtension: UITestStubLibraryService.fileExtension
        )
    }

    /// Copies the bundled sample into a fresh temporary directory so a test can mutate or delete it
    /// without touching the bundle.
    static func temporarySampleBookURL() throws -> URL {
        guard let source = sampleBookURL else {
            throw CocoaError(.fileNoSuchFile)
        }
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let destination = directory.appendingPathComponent(
            "\(UITestStubLibraryService.resourceName).\(UITestStubLibraryService.fileExtension)"
        )
        try FileManager.default.copyItem(at: source, to: destination)
        return destination
    }

    static var samplePDFURL: URL? {
        Bundle.main.url(forResource: "SampleBook", withExtension: "pdf")
    }

    static func temporarySamplePDFURL() throws -> URL {
        guard let source = samplePDFURL else {
            throw CocoaError(.fileNoSuchFile)
        }
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let destination = directory.appendingPathComponent("SampleBook.pdf")
        try FileManager.default.copyItem(at: source, to: destination)
        return destination
    }

    static func locator(
        href: String = "OEBPS/chapter1.xhtml",
        mediaType: MediaType = .xhtml,
        title: String? = "Chapter One",
        progression: Double? = 0.25,
        totalProgression: Double? = 0.1,
        position: Int? = 3,
        highlight: String? = "The ship left harbour before dawn."
    ) -> Locator {
        Locator(
            href: AnyURL(string: href)!,
            mediaType: mediaType,
            title: title,
            locations: Locator.Locations(
                progression: progression,
                totalProgression: totalProgression,
                position: position
            ),
            text: Locator.Text(highlight: highlight)
        )
    }
}

File: android/app/src/test/java/miko/biblossurfer/support/TestFixtures.kt
package miko.biblossurfer.support

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import miko.biblossurfer.data.UITestStubLibraryService
import miko.biblossurfer.data.bundledEpubFiles
import miko.biblossurfer.data.bundledPdfFile
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.Url
import org.readium.r2.shared.util.mediatype.MediaType
import java.io.File
import java.util.UUID

object TestFixtures {
    val sampleBookTitle: String = UITestStubLibraryService.sampleTitle
    val sampleBookAuthor: String = UITestStubLibraryService.sampleAuthor

    private val context: Context
        get() = ApplicationProvider.getApplicationContext()

    val sampleBookFile: File?
        get() = context.bundledEpubFiles().firstOrNull {
            it.name == "${UITestStubLibraryService.resourceName}.${UITestStubLibraryService.fileExtension}"
        }

    fun temporarySampleBookFile(): File {
        val source = sampleBookFile ?: error("sample EPUB missing")
        val directory = File(context.cacheDir, UUID.randomUUID().toString())
        directory.mkdirs()
        val destination = File(
            directory,
            "${UITestStubLibraryService.resourceName}.${UITestStubLibraryService.fileExtension}",
        )
        source.copyTo(destination)
        return destination
    }

    val samplePdfFile: File?
        get() = context.bundledPdfFile()

    fun temporarySamplePdfFile(): File {
        val source = samplePdfFile ?: error("sample PDF missing")
        val directory = File(context.cacheDir, UUID.randomUUID().toString())
        directory.mkdirs()
        val destination = File(directory, "SampleBook.pdf")
        source.copyTo(destination)
        return destination
    }

    val gutenbergHistoryFile: File?
        get() = context.bundledEpubFiles().firstOrNull { it.name.startsWith("pg28876") }

    fun locator(
        href: String = "OEBPS/chapter1.xhtml",
        mediaType: MediaType = MediaType.XHTML,
        title: String? = "Chapter One",
        progression: Double? = 0.25,
        totalProgression: Double? = 0.1,
        position: Int? = 3,
        highlight: String? = "The ship left harbour before dawn.",
    ): Locator = Locator(
        href = Url(href)!!,
        mediaType = mediaType,
        title = title,
        locations = Locator.Locations(
            progression = progression,
            totalProgression = totalProgression,
            position = position,
        ),
        text = Locator.Text(highlight = highlight),
    )
}
